parameters/parameters.go

```diff
@@ -0,0 +1,197 @@
+		parameters = append(parameters,
+			&cloudformation2.Parameter{
+				ParameterValue: &params[paramnum].ParameterValue,
+				ParameterKey:   &params[paramnum].ParameterKey})
+	}
+	return
+}
+
+func GetParameters(context *context.Context) (parameters []*Parameter, err error) {
+	templateFile, err := parseTemplate(context)
+	if err != nil {
+		context.Logger.Error(err.Error())
+		return nil, err
+	}
+	for parameterName, parameterSpec := range templateFile.Parameters {
+		var parameterValid bool
+		var parameterValue string
+		if context.CliArguments.Parameters != nil {
+			var exists bool
+			parameterValue, exists = (*context.CliArguments.Parameters)[parameterName]
+			if exists {
+				parameterValid, err = checkParameterValid(parameterName, parameterSpec.(map[string]interface{}), parameterValue, context)
+			}
+		} else {
+			parameterValid = false
+		}
+		for !parameterValid {
+			print(parameterName, ": ")
+			fmt.Scanf("%s", &parameterValue)
+			parameterValid, err = checkParameterValid(parameterName, parameterSpec.(map[string]interface{}), parameterValue, context)
+			if err != nil {
+				context.Logger.Error(err.Error())
+				return
+			}
+		}
+		parameters = append(parameters, &Parameter{ParameterKey: parameterName, ParameterValue: parameterValue})
+	}
+	return
+}
+
+func checkParameterValid(parameterName string, parameterArgument map[string]interface{}, parameterValue string, context *context.Context) (bool, error) {
+	if parameterArgument["AllowedValues"] != nil {
+		allowedValues := getAllowedValues(parameterArgument)
+		if !sliceContains(parameterValue, allowedValues) {
+			context.Logger.Error("Value '" + parameterValue + "' is not allowed for Parameter " + parameterName + ". Value must be one of following: [" + strings.Join(allowedValues, ", ") + "]")
+			return false, nil
+		}
+	}
+
+	if parameterArgument["AllowedPattern"] != nil {
+		allowedPattern := parameterArgument["AllowedPattern"].(string)
+		matches, err := regexp.Match(allowedPattern, []byte(parameterValue))
+		if err != nil {
+			return false, err
+		}
+		if !matches {
+			context.Logger.Error("Value '" + parameterValue + "' does not match the required pattern for Parameter " + parameterName)
+			return false, nil
+		}
+	}
+	return true, nil
+}
+func getAllowedValues(parameterArgument map[string]interface{}) (res []string) {
+	list := parameterArgument["AllowedValues"].([]interface{})
+	for _, val := range list {
+		res = append(res, val.(string))
+	}
+	return
+}
+
+func sliceContains(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+func parseTemplate(context *context.Context) (res cloudformation.Template, err error) {
+	rawTemplate, err := ioutil.ReadFile(*context.CliArguments.TemplatePath)
+	if err != nil {
+		return
+	}
+
+	myTemplate := template.Template{}
+
+	templateFileExtension := path.Ext(*context.CliArguments.TemplatePath)
+	if templateFileExtension == ".json" {
```

Comment:
I believe extension extraction code and decision making can be transferred into common helper, with higher order functions support.